exporter/datadogexporter/translate_traces.go

```diff
@@ -0,0 +1,492 @@
+	if serviceName == "" && cfg.Service != "" {
+		serviceName = cfg.Service
+	}
+
+	version := cfg.Version
+	tags := aggregateSpanTags(s, datadogTags)
+
+	// if no version tag exists, set it if provided via config
+	if version != "" {
+		if tagVersion := tags[versionTag]; tagVersion == "" {
+			tags[versionTag] = version
+		}
+	}
+
+	// get tracestate as just a general tag
+	if len(s.TraceState()) > 0 {
+		tags[tracetranslator.TagW3CTraceState] = string(s.TraceState())
+	}
+
+	// get start/end time to calc duration
+	startTime := s.StartTime()
+	endTime := s.EndTime()
+	duration := int64(endTime) - int64(startTime)
+
+	// it's possible end time is unset, so default to 0 rather than using a negative number
+	if s.EndTime() == 0 {
+		duration = 0
+	}
+
+	datadogType := spanKindToDatadogType(s.Kind())
+
+	span := &pb.Span{
+		TraceID:  decodeAPMId(s.TraceID().Bytes()[:]),
+		SpanID:   decodeAPMId(s.SpanID().Bytes()[:]),
+		Name:     getDatadogSpanName(s, tags),
+		Resource: getDatadogResourceName(s, tags),
+		Service:  serviceName,
+		Start:    int64(startTime),
+		Duration: int64(duration),
+		Metrics:  map[string]float64{},
+		Meta:     map[string]string{},
+		Type:     datadogType,
+	}
+
+	if len(s.ParentSpanID().Bytes()) > 0 {
+		span.ParentID = decodeAPMId(s.ParentSpanID().Bytes()[:])
+	}
+
+	// Set Span Status and any response or error details
+	status := s.Status()
+	if !status.IsNil() {
+		// map to known msg + status
+		code, ok := statusCodes[int32(status.Code())]
+		if !ok {
+			code = codeDetails{
+				message: "ERR_CODE_" + strconv.FormatInt(int64(status.Code()), 10),
+				status:  http.StatusInternalServerError,
+			}
+		}
+
+		// check if 500 or 400 level, if so set error
+		if code.status/100 == 5 || code.status/100 == 4 {
+			span.Error = 1
+			// set error type
+			tags[ext.ErrorType] = code.message
+			// set error message
+			if status.Message() != "" {
+				tags[ext.ErrorMsg] = status.Message()
+			} else {
+				tags[ext.ErrorMsg] = code.message
+			}
+			// otherwise no error
+		}
+
+		// set tag as string for status code
+		tags[httpStatusCode] = status.Code().String()
+	}
+
+	// Set Attributes as Tags
+	for key, val := range tags {
+		setStringTag(span, key, val)
+	}
+
+	for _, val := range globalTags {
+		parts := strings.Split(val, ":")
+		// only apply global tag if its not service/env/version/host and it is not malformed
+		if len(parts) < 2 || strings.TrimSpace(parts[1]) == "" || isCanonicalSpanTag(strings.TrimSpace(parts[0])) {
+			continue
+		}
+
+		setStringTag(span, strings.TrimSpace(parts[0]), strings.TrimSpace(parts[1]))
+	}
+
+	return span, nil
+}
+
+func resourceToDatadogServiceNameAndAttributeMap(
+	resource pdata.Resource,
+) (serviceName string, datadogTags map[string]string) {
+
+	datadogTags = make(map[string]string)
+	if resource.IsNil() {
+		return tracetranslator.ResourceNotSet, datadogTags
+	}
+
+	attrs := resource.Attributes()
+	if attrs.Len() == 0 {
+		return tracetranslator.ResourceNoAttrs, datadogTags
```

Comment:
Curious why this and ResourceNotSet would be preferable to `ResourceNoServiceName` as a Datadog service name.

Reply:
There's no real strong beliefs here, I think I was mostly mimicking the behavior of zipkin's translation logic, here: https://github.com/open-telemetry/opentelemetry-collector/blob/035aa5cf6c92678c3f8362d6916e18444a8250fb/translator/trace/zipkin/zipkinv2_to_traces.go#L353-L360

That being said it's less to worry about to just have a canonical "There's no resource `service.name` tag", so I've followed your advice here and just simplified to return `ResourceNoServiceName` in either case.